### PR TITLE
Add Teradata to GenericSQL integration

### PR DESCRIPTION
**Pull request by rshunim**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-7219

## Description
Added support for another DB called Teradata to GenericSQL integration.

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 1]
```diff
@@ -282,11 +291,21 @@ def sql_query_execute(client: Client, args: dict, *_) -> Tuple[str, Dict[str, An
...
+        if client.dialect == "Teradata":
+            table = []
+            for row in result:
+                convert_tup_to_dic = {header: col for header, col in zip(headers, row)}
+                table.append(convert_tup_to_dic)
+        else:
+            # converting a sqlalchemy object to a table
+            converted_table = [dict(row) for row in result]
+            # converting b'' and datetime objects to readable ones
+            table = [{str(key): str(value) for key, value in dictionary.items()} for dictionary in converted_table]
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:304]: the `dictionary` is actually a table row, right? let's call it that, to make debugging easier

Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py]: Done

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 2]
```diff
@@ -157,20 +160,26 @@ def sql_query_execute_request(self, sql_query: str, bind_vars: Any, fetch_limit=
         headers = []
         if results:
             # if the table isn't empty
-            headers = list(results[0].keys() if results[0].keys() else '')
+            if self.dialect == "Teradata":
```
[the remark this answers is not in the document]
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:169]: Done

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 3]
```diff
@@ -282,11 +291,21 @@ def sql_query_execute(client: Client, args: dict, *_) -> Tuple[str, Dict[str, An
...
         result, headers = client.sql_query_execute_request(sql_query, bind_variables)
-        # converting an sqlalchemy object to a table
-        converted_table = [dict(row) for row in result]
-        # converting b'' and datetime objects to readable ones
-        table = [{str(key): str(value) for key, value in dictionary.items()} for dictionary in converted_table]
+
+        if client.dialect == "Teradata":
+            table = []
+            for row in result:
+                convert_tup_to_dic = {header: col for header, col in zip(headers, row)}
```
[the remark this answers is not in the document]
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py]: Done

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 4]
```diff
@@ -282,11 +291,21 @@ def sql_query_execute(client: Client, args: dict, *_) -> Tuple[str, Dict[str, An
...
+                table.append(convert_tup_to_dic)
```
[the remark this answers is not in the document]
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py]: Done

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 5]
```diff
@@ -282,11 +291,21 @@ def sql_query_execute(client: Client, args: dict, *_) -> Tuple[str, Dict[str, An
         bind_variables = generate_bind_vars(bind_variables_names, bind_variables_values)
 
         result, headers = client.sql_query_execute_request(sql_query, bind_variables)
-        # converting an sqlalchemy object to a table
-        converted_table = [dict(row) for row in result]
-        # converting b'' and datetime objects to readable ones
-        table = [{str(key): str(value) for key, value in dictionary.items()} for dictionary in converted_table]
+
+        if client.dialect == "Teradata":
```
[the remark this answers is not in the document]
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py]: Done

**Comment by ShirleyDenkberg** [comment]
@dorschw Doc review completed.

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 6]
```diff
@@ -129,12 +133,21 @@ def _create_engine_and_connect(self) -> sqlalchemy.engine.base.Connection:
...
                 ssl_connection = {'ssl': {'ssl-mode': 'preferred'}}  # type: ignore[dict-item]
-        engine: sqlalchemy.engine.Engine = None
+
+        # Teradata has a unique connection, unlike the others with URL object
+        if self.dialect == TERADATA:
+            if self.use_ldap:
+                db_url = f'teradatasql://{self.username}:{self.password}@{self.host}:{self.port}/?logmech=LDAP'
+            else:
+                db_url = f'teradatasql://{self.host}:{self.port}/?user={self.username}&password={self.password}'
+            demisto.debug('Initializing engine using the Teradata dialect')
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:149]: let's put this, together with `db_url = URL(...)`, in their own method called `_generate_db_url`

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 7]
```diff
@@ -157,20 +160,26 @@ def sql_query_execute_request(self, sql_query: str, bind_vars: Any, fetch_limit=
         headers = []
         if results:
             # if the table isn't empty
-            headers = list(results[0].keys() if results[0].keys() else '')
+            if self.dialect == "Teradata":
+                headers = list(result.keys())
```
[the remark this answers is not in the document]
Comment (dorschw) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:192]: where? :)

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 8]
```diff
@@ -317,6 +336,23 @@ def test_module(client: Client, *_) -> tuple[str, dict[Any, Any], list[Any]]:
...
+    """
+    if client.dialect == TERADATA:
+        # binding the headers with the columns
+        converted_table = [dict(zip(headers, row)) for row in result]
+    else:
+        # converting a sqlalchemy object to a table
+        converted_table = [dict(row) for row in result]
+
+    # converting b'' and datetime objects to readable ones
+    table = [{str(key): str(value) for key, value in dictionary.items()} for dictionary in converted_table]
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:351]: is this unit-tested?

Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:351]: Does this only apply for Teradata?

Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:351]: Are we sure this won't break complex values (binary, timestamp) in existing instances on update?

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py [hunk 9]
```diff
@@ -129,12 +133,21 @@ def _create_engine_and_connect(self) -> sqlalchemy.engine.base.Connection:
...
             if self.dialect == POSTGRES_SQL:
                 ssl_connection = {'sslmode': 'require'}
             else:
                 ssl_connection = {'ssl': {'ssl-mode': 'preferred'}}  # type: ignore[dict-item]
-        engine: sqlalchemy.engine.Engine = None
+
+        # Teradata has a unique connection, unlike the others with URL object
+        if self.dialect == TERADATA:
+            if self.use_ldap:
+                db_url = f'teradatasql://{self.username}:{self.password}@{self.host}:{self.port}/?logmech=LDAP'
```
Comment (dorschw) [inline Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:146]: let's discuss this offline

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py (hunk 6 above)
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:149]: Done

File: Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py (hunk 8 above)
Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:351]: No, for all DBs

Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:351]: Yes, we sure. This line is not a new line, just in a new place in the code.

Comment (rshunim) [reply Packs/GenericSQL/Integrations/GenericSQL/GenericSQL.py:351]: Yes

**Review by dorschw — APPROVED** [review]

**Pull request merged**